messaging/src/main/java/org/axonframework/commandhandling/gateway/AbstractRetryScheduler.java

```diff
@@ -146,6 +173,106 @@ public boolean scheduleRetry(CommandMessage commandMessage,
+            assertNonNull(failureType, "Class of failure type may not be null");
+            assertNonNull(nonTransientFailurePredicate, "Predicate may not be null");
+
+            //noinspection Convert2MethodRef
+            Predicate<E> typeCheckPredicate = (failureAtRuntime) -> failureType.isInstance(failureAtRuntime);
+
+            // noinspection unchecked
+            this.nonTransientFailurePredicate = ((Predicate<Throwable>)typeCheckPredicate.and(nonTransientFailurePredicate));
+
+            // noinspection unchecked
+            return (B) this;
+        }
+
+        /**
+         * Adds additional predicate into the predicate chain for checking non-transiency of a failure.
+         * <p/>
+         * Provided {@code nonTransientFailurePredicate} is set on the beginning of the predicate chain with
+         * {@code or} operator.
+         *
+         * @param nonTransientFailurePredicate an additional {@link Predicate} (accepting a {@link Throwable}
+         *                                     parameter) to be added into the predicate chain used for
+         *                                     testing transiency of provided failure
+         * @return the current Builder instance, for fluent interfacing
+         */
+        public B addNonTransientFailurePredicate(Predicate<Throwable> nonTransientFailurePredicate) {
+            assertNonNull(nonTransientFailurePredicate, "Predicate may not be null");
+            this.nonTransientFailurePredicate = nonTransientFailurePredicate.or(this.nonTransientFailurePredicate);
+
+            // noinspection unchecked
+            return (B) this;
+        }
+
+        /**
+         * Adds additional predicate into the predicate chain for checking non-transiency of a failure.
+         * <p/>
+         * Provided {@code nonTransientFailurePredicate} is set on the beginning of the predicate chain with
+         * {@code or} operator.
+         * <p/>
+         * In essence, this is just a failure typed variant of {@link #addNonTransientFailurePredicate(Predicate)}
+         * method.
+         *
+         * @param failureType concrete class of the failure to handle
+         * @param nonTransientFailurePredicate an additional {@link Predicate} (accepting a {@code failure}
+         *                                     parameter) to be added into the predicate chain used for testing
+         *                                     transiency of provided failure
+         * @param <E> type of the failure to handle
+         * @return the current Builder instance, for fluent interfacing
+         */
+        public <E extends Throwable> B addNonTransientFailurePredicate(Class<E> failureType, Predicate<? super E> nonTransientFailurePredicate) {
+            assertNonNull(failureType, "Class of failure type may not be null");
+            assertNonNull(nonTransientFailurePredicate, "Predicate may not be null");
```

Comment:
I'd be a little more specific in the exception message, stating that the `"Non-transient failure predicate may not be null"`